Explain API for Exact/ANN/Radial/Disk based KNN search on Faiss

neetikasinghal opened the pull request:
### Description
Add support for explain for Exact/ANN/Radial/Disk/Filtering k-nn search. Score calculation explanation is currently added only for ANN search.
Proposal for explain is given here: https://github.com/opensearch-project/k-NN/issues/875#issuecomment-2611349466

### Related Issues
Resolves #875
<!-- List any other related issues here -->

### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Vikasht34 commented on src/main/java/org/opensearch/knn/index/query/KNNWeight.java:
```diff
@@ -106,7 +110,100 @@ static void initialize(ModelDao modelDao, ExactSearcher exactSearcher) {
 
     @Override
     public Explanation explain(LeafReaderContext context, int doc) {
```
Could we improve readability of this function , I see this function is doing 

1. Building Score 
2. Build the main explanation
3. Build the detailed explanation for the leaf node.

Would we start with something simple 

```
public Explanation explain(LeafReaderContext context, int doc) {
   //Function to Calculate Score
    // Build the main explanation

      // Build the detailed explanation for the leaf node
    return Explanation.match(score, mainExplanation, Explanation.match(score, leafExplanation));
}
```
And Please break down other functions as well when needed.

neetikasinghal replied:
sure i will refactor this.

Vikasht34 reviewed (commented):
We need to have Unit test for each case 
1. Disk-based search with valid rescore context.
2. Radial search.
3. ANN search (default case).
4. Shard-level rescoring enabled.
5. Shard-level rescoring disabled.
6.  Filter weight case where filtered IDs are less than k.
7. Filter threshold value greater than cardinality.
8. Missing native engine files.
9. Valid context with matching document and disk-based search.

And Please validate with Explaination Object.

neetikasinghal commented:
> We need to have Unit test for each case
> 
> 1. Disk-based search with valid rescore context.
> 2. Radial search.
> 3. ANN search (default case).
> 4. Shard-level rescoring enabled.
> 5. Shard-level rescoring disabled.
> 6. Filter weight case where filtered IDs are less than k.
> 7. Filter threshold value greater than cardinality.
> 8. Missing native engine files.
> 9. Valid context with matching document and disk-based search.
> 
> And Please validate with Explaination Object.

@Vikasht34 yes the tests are not yet added in here, hence its in a draft status. I will add the coverage with all the possible cases.

neetikasinghal commented:
@navneet1v / @Vikasht34 would you please review the changes?

navneet1v commented:
@neetikasinghal can you please add an entry in the change log

neetikasinghal commented:
> @neetikasinghal can you please add an entry in the change log

yup i generally add it towards the end of the review so that its easier to rebase with the latest changes.

navneet1v commented on src/main/java/org/opensearch/knn/index/SpaceType.java:
```diff
@@ -46,6 +51,11 @@ public float scoreTranslation(float rawScore) {
 
+        @Override
+        public String explainScoreTranslation(float rawScore) {
+            return "`1 / (1 + rawScore)`";
```
can we make this as private static final string.

navneet1v commented on src/main/java/org/opensearch/knn/index/SpaceType.java:
```diff
@@ -77,6 +87,11 @@ public float scoreTranslation(float rawScore) {
 
+        @Override
+        public String explainScoreTranslation(float rawScore) {
+            return "`Math.max((2.0F - rawScore) / 2.0F, 0.0F)`";
```
same as above

navneet1v commented on src/main/java/org/opensearch/knn/index/query/KNNWeight.java:
```diff
@@ -105,8 +109,147 @@ static void initialize(ModelDao modelDao, ExactSearcher exactSearcher) {
+            } catch (IOException e) {
+                throw new RuntimeException("Error while getting KNN score during explanation", e);
+            }
+            if (score == 0) {
+                score = knnScore;
+            }
```
shouldn't this be the first check we do when we enter the if condition.

neetikasinghal replied:
not really, knnScore should be computed for it to be assigned to the score.

navneet1v commented on src/main/java/org/opensearch/knn/index/query/KNNWeight.java:
```diff
@@ -105,8 +109,147 @@ static void initialize(ModelDao modelDao, ExactSearcher exactSearcher) {
+    }
+
+    private float getKnnScore(LeafReaderContext context, int doc) throws IOException {
+        KNNScorer knnScorer = (KNNScorer) scorer(context);
```
why we are computing the scorer for every doc again and again? In case of explain should we just cache scorer in the KNNExplainObject once the first explain call happens.

neetikasinghal replied:
thats a good point, will make changes

navneet1v reviewed (commented):
Looked at high level in the code, need to go through more in terms of explanation. But one thing I want to add is I am not seeing any ITs related to explain api. Can we please add them too

neetikasinghal commented:
> Looked at high level in the code, need to go through more in terms of explanation. But one thing I want to add is I am not seeing any ITs related to explain api. Can we please add them too

yup, that WIP as my setup for ITs was broken. I am able to fix that now, however the PR has the coverage for all the UTs.

Vikasht34 reviewed (commented):
Will continue on other files.

Vikasht34 commented on src/main/java/org/opensearch/knn/index/query/explain/KnnExplanation.java:
```diff
@@ -0,0 +1,40 @@
+ * This class captures details around knn explain queries that is used
+ * by explain API to generate explanation for knn queries
+ */
+public class KnnExplanation {
```
There are couple of things that we can improve on this class, 

1.  As of Now every client of this class has to get Maps and then put values so Maps are final but Modificable outside of this class. Ideally as  client,  I should not be knowing internal data structure of KnnExplanation Class ,  Client must to exposed to api like addAnnResult, AddRawScores , addKnnScores that's all , how we are adding these and which data structure to use , it's should be upto KNNExplanation class. 
2. Please make all maps Unmodifiable outside of this class. No Client should be able to Modify these maps except KNNExplanation.

neetikasinghal replied:
makes sense, will update it.

Vikasht34 commented on src/main/java/org/opensearch/knn/index/query/common/DocAndScoreQuery.java:
```diff
@@ -31,13 +32,15 @@ final class DocAndScoreQuery extends Query {
+    private final KNNWeight knnWeight;
 
-    public DocAndScoreQuery(int k, int[] docs, float[] scores, int[] segmentStarts, Object contextIdentity) {
+    public DocAndScoreQuery(int k, int[] docs, float[] scores, int[] segmentStarts, Object contextIdentity, KNNWeight knnWeight) {
```
Now whole explain method is there in KNNWeight class , DocAndScoreQuery has to take reference of this class , just to call , explain method of this class , Please abstract explain method from KnnWeight class so that in future if we have another type of query , we don't end up passing KnnWeight class everywhere.

neetikasinghal replied:
i understand the caveat here of passing the KNNWeight class, but didnt really get your suggestion of abstracting out the explain method out of the KNNWeight class.

Vikasht34 replied:
Please move explain method to utlity class , which takes context , k and scorere and call that utility method from both KnnWieght and DocsAndScoREqUERY

neetikasinghal replied:
so KnnWeight has access to KnnQuery which is needed as part of explain function determining the explanation, knnQuery is not present in DocAndScoreQuery. Also, we are setting the explain flag as true as part of knnQuery to set some maps in the searchLeaf path which would not be set if the explain is set to false.

@Vikasht34 let me know if this makes sense or if u have a better way of doing it.

Vikasht34 replied:
I think that's fine .

neetikasinghal commented on src/main/java/org/opensearch/knn/index/SpaceType.java:
```diff
@@ -177,6 +215,8 @@ public KNNVectorSimilarityFunction getKnnVectorSimilarityFunction() {
 
     public abstract float scoreTranslation(float rawScore);
 
+    public abstract String explainScoreTranslation(float rawScore);
```
i can introduce introduce the static final variables to reduce code duplication, however the logic highlighted above of constructors wont work in case of inner product where the explain formula differs:
```
@Override
        public String explainScoreTranslation(float rawScore) {
            if (rawScore >= 0) {
                return "`1 / (1 + rawScore)`";
            }
            return "`-rawScore + 1`";
        }
```

Its dependent on raw score. Also, there could be future space type whose formula depends on the raw score, hence i thought of introducing this function instead. let me know your thoughts here.

Vikasht34 replied:
This is how Ineer product will work, if explainScoreTranslation is differnt then it can be overridden.

```
INNER_PRODUCT("innerproduct") {
        @Override
        public float scoreTranslation(float rawScore) {
            return rawScore >= 0 ? (1 / (1 + rawScore)) : (-rawScore + 1);
        }

        @Override
        public String explainScoreTranslation(float rawScore) {
            return rawScore >= 0 ? "`1 / (1 + rawScore)`" : "`-rawScore + 1`";
        
```

Vikasht34 replied:
```
public enum SpaceType {
    UNDEFINED("undefined") {
        @Override
        public float scoreTranslation(final float rawScore) {
            throw new IllegalStateException("Unsupported method");
        }

        @Override
        public String explainScoreTranslation(float rawScore) {
            throw new IllegalStateException("Unsupported method");
        }
    },

    L2("l2", "`1 / (1 + rawScore)`") {
        @Override
        public float scoreTranslation(float rawScore) {
            return 1 / (1 + rawScore);
        }
    },

    COSINE("cosine", "`Math.max((2.0F - rawScore) / 2.0F, 0.0F)`") {
        @Override
        public float scoreTranslation(float rawScore) {
            return Math.max((2.0F - rawScore) / 2.0F, 0.0F);
        }
    },

    INNER_PRODUCT("innerproduct") {
        @Override
        public float scoreTranslation(float rawScore) {
            return rawScore >= 0 ? (1 / (1 + rawScore)) : (-rawScore + 1);
        }

        @Override
        public String explainScoreTranslation(float rawScore) {
            return rawScore >= 0 ? "`1 / (1 + rawScore)`" : "`-rawScore + 1`";
        }
    };

    private final String name;
    private final String explanationFormula;

    SpaceType(String name) {
        this.name = name;
        this.explanationFormula = null;
    }

    SpaceType(String name, String explanationFormula) {
        this.name = name;
        this.explanationFormula = explanationFormula;
    }

    public abstract float scoreTranslation(float rawScore);

    public String explainScoreTranslation(float rawScore) {
        if (explanationFormula != null) {
            return explanationFormula;
        }
        throw new UnsupportedOperationException("explainScoreTranslation not defined for this space type.");
    }
}

```

Let me know if it makes sense.

neetikasinghal replied:
yes this makes sense, will update it.

neetikasinghal commented:
pending items -
- rebase to the latest changes
- Add ITs

martin-gaievski commented on src/main/java/org/opensearch/knn/index/query/KNNWeight.java:
```diff
@@ -105,8 +109,153 @@ static void initialize(ModelDao modelDao, ExactSearcher exactSearcher) {
+    public Explanation explain(LeafReaderContext context, int doc, float score, KNNScorer knnScorer) {
+        knnQuery.setExplain(true);
+        try {
+            knnScorer = getOrCreateKnnScorer(context, knnScorer);
```
is there a scenario when knnScorer is not null? If no, then you should not pass it to this method and always create new instance inside the method.

neetikasinghal replied:
makes sense, removed the knnScorer being passed in the function.

martin-gaievski commented on src/main/java/org/opensearch/knn/index/query/KNNWeight.java:
```diff
@@ -105,8 +109,153 @@ static void initialize(ModelDao modelDao, ExactSearcher exactSearcher) {
+        try {
+            return FieldInfoExtractor.getSpaceType(modelDao, fieldInfo);
+        } catch (IllegalArgumentException e) {
+            return knnQuery.getVectorDataType() == VectorDataType.BINARY ? SpaceType.DEFAULT_BINARY : SpaceType.DEFAULT;
```
why we are not signaling to user that system is in bad state, wouldn't this be misleading, I mean giving false data to the user?

neetikasinghal replied:
makes sense, i think we dont need to handle this exception here, will remove this. It should be taken care of during index creation itself.

martin-gaievski commented on src/main/java/org/opensearch/knn/index/SpaceType.java:
```diff
@@ -169,14 +174,29 @@ public KNNVectorSimilarityFunction getKnnVectorSimilarityFunction() {
+        if (explanationFormula != null) {
+            return explanationFormula;
+        }
+        throw new UnsupportedOperationException("explainScoreTranslation is not defined for this space type.");
```
this effectively means if object created with old constructor (formula is null) then this method will throw exception, is this desired behavior? Can we have some constant like "undefined" or similar, or if formula isn't defined the rets of explanation detail don't make sense as well?

neetikasinghal replied:
@martin-gaievski The old constructor is called for two enums: `UNDEFINED` and `INNER_PRODUCT`. `INNER_PRODUCT` overrides the `explainScoreTranslation` method and for undefined its fine to throw an exception.
The callers of this function would know that the explainScoreTranslation is not defined for a specific space type. I think that should be fine, what do u think?

martin-gaievski commented on src/main/java/org/opensearch/knn/index/query/explain/KnnExplanation.java:
```diff
@@ -0,0 +1,56 @@
+public class KnnExplanation {
+
+    @Getter
+    private final Map<Object, Integer> annResultPerLeaf;
```
are you sure you want to expose raw map? clients will be able to modify them. You may want to consider returning `Collections.unmodifiableMap(your_map)` in getter method

neetikasinghal replied:
good catch! removed the getter for this map as well as other maps in the class too.

neetikasinghal commented:
> pending items -
> 
> * rebase to the latest changes
> * Add ITs

completed the pending items and addressed all the comments.

@navneet1v @Vikasht34 @martin-gaievski can u please take another look at it?

Vikasht34 reviewed (approved):
LGTM , Thanks for addressing the suggestions !!

navneet1v commented on CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased 2.x](https://github.com/opensearch-project/k-NN/compare/2.19...2.x)
 ### Features
+* [Explain API for Exact/ANN/Radial/Disk based KNN search on Faiss] Explain API for Exact/ANN/Radial/Disk based KNN search on Faiss [#2403] (https://github.com/opensearch-project/k-NN/pull/2403)
```
```suggestion
* [Explain API Support] Added Explain API support for Exact/ANN/Radial/Disk based KNN search on Faiss Engine [#2403] (https://github.com/opensearch-project/k-NN/pull/2403)
```

navneet1v commented on src/main/java/org/opensearch/knn/index/SpaceType.java:
```diff
@@ -169,14 +174,29 @@ public KNNVectorSimilarityFunction getKnnVectorSimilarityFunction() {
 
     public abstract float scoreTranslation(float rawScore);
 
+    public String explainScoreTranslation(float rawScore) {
+        if (explanationFormula != null) {
+            return explanationFormula;
+        }
```
what is the use of rawScore parameter here when that parameter is not getting used at all?

neetikasinghal replied:
its used for a space type like `INNER_PRODUCT,` check: https://github.com/opensearch-project/k-NN/pull/2403/files#diff-c481eefa59727056eb79ce9614cbcbb8b84358cf4c8b4bc51eb2d089b8a60a1bR132

navneet1v commented on src/main/java/org/opensearch/knn/index/query/KNNWeight.java:
```diff
@@ -108,9 +112,148 @@ static void initialize(ModelDao modelDao, ExactSearcher exactSearcher) {
+                score = getKnnScore(knnScorer, doc);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Error while explaining KNN score", e);
```
```suggestion
            throw new RuntimeException("Error while explaining KNN score for doc: " + doc + " score: " + score, e);
```

navneet1v commented on src/main/java/org/opensearch/knn/index/query/KNNWeight.java:
```diff
@@ -108,9 +112,148 @@ static void initialize(ModelDao modelDao, ExactSearcher exactSearcher) {
+        knnQuery.setExplain(true);
+        try {
+            final KNNScorer knnScorer = getOrCreateKnnScorer(context);
+            if (score == 0) {
+                score = getKnnScore(knnScorer, doc);
+            }
```
why this condition and what will happen if score is not zero?

neetikasinghal replied:
if the score is 0, it will calculate the score, however if score is passed from the caller, like in case of disk based search, there is no need to re-compute the score and we can rely on the score passed from the caller.

navneet1v replied:
It would be better to provide this information as a Java doc.

neetikasinghal replied:
sure i can add it

navneet1v commented on src/main/java/org/opensearch/knn/index/query/KNNWeight.java:
```diff
@@ -108,9 +112,148 @@ static void initialize(ModelDao modelDao, ExactSearcher exactSearcher) {
+    }
+
+    private String buildDiskBasedSearchExplanation() {
+        StringBuilder sb = new StringBuilder(KNNConstants.DISK_BASED_SEARCH);
+        boolean isShardLevelRescoringDisabled = KNNSettings.isShardLevelRescoringDisabledForDiskBasedVector(knnQuery.getIndexName());
+        if (!knnQuery.getRescoreContext().isRescoreEnabled()) {
+            isShardLevelRescoringDisabled = true;
+        }
```
why the disk based rescore logic coming in this query? this logic in explain should be part of NativeEngineKnnVectorQuery

neetikasinghal replied:
NativeEngineKnnVectorQuery is not aware of any explain related details, it flows down to KNNWeight.

navneet1v replied:
I understand the part that we use KNNWeight for both first pass and second pass in disk based vector search. But the decision on wether to use the shard level or segment level rescoring is not controlled by KNNWeight. Hence this is just extra information which kNNWeight has to bear now due to explain API, which in my true sense is not a good design.

neetikasinghal replied:
the problem is NativeEngineKnnVectorQuery doesnt handle all the search use-cases, for Radial search, the logic is handled separately which also looks to be redundant. If we can have all the logic in `NativeEngineKnnVectorQuery`, i think we can move the explanation part to the `NativeEngineKnnVectorQuery` as well. 
KNNWeight is not bearing this extra information, its just one check that it has to do in case of disk-based search.
I think we should refactor radial search also to have the same code flow as `NativeEngineKnnVectorQuery`.

Let me know your thoughts and if u have any other suggestions.

neetikasinghal replied:
created issue for this: https://github.com/opensearch-project/k-NN/issues/2670

navneet1v commented on src/test/java/org/opensearch/knn/index/ExplainIT.java:
```diff
@@ -0,0 +1,333 @@
+public class ExplainIT extends KNNRestTestCase {
+
+    @SneakyThrows
+    public void testAnnSearch() {
```
Please rename all test functions to `testABC_whenDEF_thenXYZ()` format to ensure better readability.

neetikasinghal replied:
i think the tests are named readable even now, i dont see a need to rename the tests unless there is a strong opinion on it.

neetikasinghal replied:
ok i can rename them.

navneet1v commented on src/test/java/org/opensearch/knn/index/ExplainIT.java:
```diff
@@ -0,0 +1,333 @@
+        );
+    }
+
+    private void validateExplainSearchResponse(XContentBuilder queryBuilder, String... descriptions) throws IOException, ParseException {
+        String responseBody = EntityUtils.toString(performSearch(INDEX_NAME, queryBuilder.toString(), "explain=true").getEntity());
+        List<Object> searchResponseHits = parseSearchResponseHits(responseBody);
+        searchResponseHits.stream().forEach(hit -> {
+            Map<String, Object> hitMap = (Map<String, Object>) hit;
+            Double score = (Double) hitMap.get("_score");
+            String explanation = hitMap.get("_explanation").toString();
+            assertNotNull(explanation);
+            for (String description : descriptions) {
+                assertTrue(explanation.contains(description));
+            }
+            assertTrue(explanation.contains(String.valueOf(score)));
+        });
+    }
```
this is not a strong assertion as these assertions are just validating that explanation is present. Please validate the contents of the explain api response too to ensure that correct strings are returned.

neetikasinghal replied:
the individual tests are passing all the description types that it is expecting in the explain response, please refer to individual tests to check what all we are asserting for.

navneet1v replied:
Got it, I think I saw what you have done. It is not validating the full string but parts of the string. `assertTrue(explanation.contains(description));`. This is still not 100% validation. We should have at-least some test which checks the full string rather than partial string.

neetikasinghal replied:
i dont agree here as the exact language can change in future and the tests would start to break which should not be the case. the current tests are validating the required description which should not change in the future.

neetikasinghal commented:
thanks @Vikasht34 for the approval. need to merge changes by 9m to make it to the cutoff for 3.0.0-beta.

would appreciate if i can get a second approval on this.

navneet1v commented:
@neetikasinghal see if you want to followup with another PR with some of the suggestion being provided as the PR is already merged.

neetikasinghal commented:
> @neetikasinghal see if you want to followup with another PR with some of the suggestion being provided as the PR is already merged.

@navneet1v had to merge this in before to make it to 3.0, created another pr for some minor comments: https://github.com/opensearch-project/k-NN/pull/2669
